csrc/preseg_passes/fmin_fmax_promotion.cpp

```diff
@@ -0,0 +1,352 @@
+//    be promoted.
+
+using NanStatusMap = std::unordered_map<TensorView*, NanStatus>;
+using PromotedOpSet = std::unordered_set<ReductionOp*>;
+
+bool isSafeReduction(Expr* expr, const PromotedOpSet& promotedOps) {
+  if (auto* rop = dynamic_cast<ReductionOp*>(expr)) {
+    // Check that this expr hasn't already been promoted to an unsafe reduction.
+    return !promotedOps.contains(rop);
+  }
+
+  return false;
+}
+
+bool reductionMatches(ReductionOp* left, ReductionOp* right) {
+  auto* left_tv = dynamic_cast<TensorView*>(left->output(0));
+  auto* right_tv = dynamic_cast<TensorView*>(right->output(0));
+
+  if (left_tv->nDims() != right_tv->nDims()) {
+    return false;
+  }
+
+  for (int i = 0; i < left_tv->nDims(); ++i) {
+    if (left_tv->getLogicalDomain()[i]->isReduction() !=
+        right_tv->getLogicalDomain()[i]->isReduction()) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool broadcastMatches(BroadcastOp* left, BroadcastOp* right) {
+  auto* left_tv = dynamic_cast<TensorView*>(left->output(0));
+  auto* right_tv = dynamic_cast<TensorView*>(right->output(0));
+
+  if (left_tv->nDims() != right_tv->nDims()) {
+    return false;
+  }
+
+  for (int i = 0; i < left_tv->nDims(); ++i) {
+    if (left_tv->getLogicalDomain()[i]->isBroadcast() !=
+        right_tv->getLogicalDomain()[i]->isBroadcast()) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool canBeAnalyzed(
+    Expr* expr,
+    ReductionOp* compareRop,
+    std::optional<BroadcastOp*>& compareBop) {
+  // This is where we enforce the restricted-subgraph rules. Arbitrary binary
+  // and binary ops are allowed, and do not affect the analysis. Reduction and
+  // broadcasts have strict requirements to simplify the state tracking.
+
+  if (expr->isA<UnaryOp>() || expr->isA<BinaryOp>()) {
+    return true;
+  } else if (auto* rop = dynamic_cast<ReductionOp*>(expr)) {
+    // We require all reduction ops exactly match in reduction axes.
+    // This avoids the need for complicated IterDomain handling.
+    return reductionMatches(rop, compareRop);
+  } else if (auto* bop = dynamic_cast<BroadcastOp*>(expr)) {
+    // Similarly for reductions, we require all broadcasts to have the same
+    // axes.
+    if (!compareBop) {
+      compareBop = bop;
+      return true;
+    } else {
+      return broadcastMatches(bop, *compareBop);
+    }
+  }
+
+  return false;
+}
+
+// Traverses the restricted subgraph around the target rop and checks whether
+// NANs which would be squelched by a promotion, will be subsequently repaired
+// by safe reductions.
+bool minMaxOpIsRepaired(
+    ReductionOp* targetRop,
+    const PromotedOpSet& promotedOps) {
+  Fusion* fusion = targetRop->fusion();
+
+  auto* in_tv = targetRop->input(0)->as<TensorView>();
+  auto* out_tv = targetRop->output(0)->as<TensorView>();
+
+  NanStatusMap status_map;
+
+  status_map.emplace(in_tv, NanStatus::Unreduced);
+  status_map.emplace(out_tv, NanStatus::BadReduced);
```

Comment:
@jacobhinkle regarding the question 

> Why not have an initialization step where all IDs of fusion inputs are marked GOOD instead of NONE?

That is kind of what we are doing here when we insert the `Unreduced` and `BadReduced` states.

Why don't we do this for all fusion inputs? Because we only analyze a subgraph of a the fusion starting at targetRop->input(0).

The reason we only work on this subgraph is because we only support status mapping for a small number of operators. If we seed data on fusion inputs, it's unlikely any meaningful status will survive up to the point where we start using it.